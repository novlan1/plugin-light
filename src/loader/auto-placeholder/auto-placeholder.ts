/**
 * 记录每个页面的动态组件
 */
export function autoPlaceLoader(source = '') {
  if (process.env.VUE_APP_PLATFORM !== 'mp-weixin' && process.env.VUE_APP_PLATFORM !== 'mp-qq') {
    return source;
  }
  const reg = new RegExp(/(([a-zA-Z]+?)\(resolve\)(?:\s*?)\{(?:\n\s*)require\(\['(.*?)'\],(?:\s*?)resolve\);(?:\n\s*)\})+/, 'g');
  const match = [...source.matchAll(reg)];
  if (!match?.length) {
    return source;
  }
  const compList = match.map((item) => {
    const name = item?.[2];
    let filePath = '';
    if (item?.[3]) {
      filePath = item[3];
      if (filePath.startsWith('src/')) {
        filePath = filePath.replace('src/', '/');
      } else if (filePath.startsWith('./')) {
        filePath = filePath.replace('./', '/');
      } else if (filePath.startsWith('../../')) {
        filePath = filePath.replace('../../', '/');
      } else if (filePath.startsWith('../')) {
        filePath = filePath.replace('../', '/');
      }
      if (filePath.endsWith('.vue')) {
        filePath = filePath.replace('.vue', '');
      }
    }
    return {
      name,
      filePath,
    };
  });
  if (!global?.placeholderMap) {
    global.placeholderMap = new Map();
  }
  // @ts-ignore
  global.placeholderMap.set(this.resourcePath.replace('.vue', ''), compList);
  return source;
}


