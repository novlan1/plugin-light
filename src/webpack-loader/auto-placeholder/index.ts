export { autoPlaceLoader as default } from './auto-placeholder';
