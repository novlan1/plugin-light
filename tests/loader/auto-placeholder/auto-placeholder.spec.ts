import path from 'path';
import { autoPlaceLoader } from '../../../src/webpack-loader/auto-placeholder/auto-placeholder';


const resourcePath = path.resolve(process.cwd(), './src', 'MOCK_PAGE.vue');

describe('autoPlaceLoader', () => {
  const oldEnv = process.env;

  beforeEach(() => {
    process.env = { ...oldEnv };
  });

  afterEach(() => {
    process.env = oldEnv;
  });

  it('not in mp-weixin or mp-qq', () => {
    const source = `
export default {
  components: {
    Card(resolve) {
      require(['./components/card'], resolve);
    },
  },
  props: {},
  data() {
    return {}
  },
  methods: {},
}`;
    const result = autoPlaceLoader.call({ resourcePath }, source);
    expect(global.placeholderMap?.get(resourcePath.replace('.vue', '')) ?? []).toEqual([]);
    expect(result).toBe(source);
  });

  it('empty components', () => {
    process.env.VUE_APP_PLATFORM = 'mp-qq';
    const source = `
export default {
  components: {},
  props: {},
  data() {
    return {}
  },
  methods: {},
}`;
    const result =  autoPlaceLoader.call({ resourcePath }, source);
    expect(global.placeholderMap?.get(resourcePath.replace('.vue', '')) ?? []).toEqual([]);
    expect(result).toBe(source);
  });

  it('with components', () => {
    process.env.VUE_APP_PLATFORM = 'mp-weixin';
    const source = `
export default {
  components: {
    Popup(resolve) {
      require(['src/components/popup'], resolve);
    },
    Dialog(resolve) {
      require(['./components/dialog.vue'], resolve);
    },
    List(resolve) {
      require(['../local-components/list'], resolve);
    },
    Item(resolve) {
      require(['../../local-components/item.vue'], resolve);
    },
  },
  props: {},
  data() {
    return {}
  },
  methods: {},
}`;
    const result =  autoPlaceLoader.call({ resourcePath }, source);
    expect(global.placeholderMap?.get(resourcePath.replace('.vue', '')) ?? []).toMatchSnapshot();
    expect(result).toBe(source);
  });
});
