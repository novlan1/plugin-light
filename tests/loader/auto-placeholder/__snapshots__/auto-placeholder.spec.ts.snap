// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autoPlaceLoader with components 1`] = `
[
  {
    "filePath": "/components/popup",
    "name": "Popup",
  },
  {
    "filePath": "/components/dialog",
    "name": "Dialog",
  },
  {
    "filePath": "/local-components/list",
    "name": "List",
  },
  {
    "filePath": "/local-components/item",
    "name": "Item",
  },
]
`;
